File: aoc2020/day2/part1.js
import fs from "fs";

const main: any = () => {
  try {
    const data: string[] = fs.readFileSync("input.txt", "utf-8").split("\n");
    let valid: number = 0;
    data.forEach((passwd) => {
      const args = passwd.split(":");
      const item1 = args[0].split(" ");
      const pass = args[1];
      const range = item1[0].split("-");
      const low = parseInt(range[0]);
      const high = parseInt(range[1]);
      const key = item1[1];

      let count: number = 0;
      for (let i = 0; i < pass.length; i++) {
        if (pass[i] === key) {
          count += 1;
        }
      }
      if (count >= low && count <= high) {
        valid += 1;
      }
    });
    console.log(valid);
  } catch (err) {
    console.error(err);
  }
};

main();

File: aoc2020/day2/part1.py
def validate_passwords():
    try:
        with open('./input.txt', 'r') as file:
            items = file.read().splitlines()
    except FileNotFoundError:
        print("Error: input.txt file not found")
        return
    except Exception as e:
        print(f"Error reading file: {e}")
        return

    count = 0
    for password_line in items:
        # Skip empty lines
        if not password_line:
            continue
            
        # Split into policy and password
        parts = password_line.split(":")
        if len(parts) != 2:
            continue

        policy, password = parts
        
        # Parse policy
        policy_parts = policy.split()
        if len(policy_parts) != 2:
            continue
            
        range_str, key = policy_parts
        
        # Parse range
        range_parts = range_str.split("-")
        if len(range_parts) != 2:
            continue
            
        try:
            low, upper = map(int, range_parts)
        except ValueError:
            continue

        # Count character occurrences
        password = password.strip()
        char_count = password.count(key)
        
        # Check if password meets policy
        if low <= char_count <= upper:
            count += 1

    print(count)

if __name__ == "__main__":
   validate_passwords()